File: app/src/main/java/com/sofps/inspirationalquotes/data/source/QuotesRepository.java
package com.sofps.inspirationalquotes.data.source;

import android.support.annotation.NonNull;
import com.sofps.inspirationalquotes.asynctask.QuotesLoader;
import com.sofps.inspirationalquotes.data.Quote;
import com.sofps.inspirationalquotes.data.source.local.QuotesLocalDataSource;
import com.sofps.inspirationalquotes.data.source.remote.QuotesRemoteDataSource;
import java.util.ArrayList;
import java.util.List;

public class QuotesRepository {

    private final QuotesLocalDataSource mQuotesLocalDataSource;
    private final QuotesRemoteDataSource mQuotesRemoteDataSource;

    public QuotesRepository(
            @NonNull final QuotesLocalDataSource quotesLocalDataSource,
            @NonNull final QuotesRemoteDataSource quotesRemoteDataSource) {
        mQuotesLocalDataSource = quotesLocalDataSource;
        mQuotesRemoteDataSource = quotesRemoteDataSource;
    }

    public void loadQuotesForLanguage(final String language, final QuotesLoader.QuotesLoaderTaskListener listener) {
        mQuotesRemoteDataSource.getQuotes(language, new QuotesDataSource.GetQuoteCallback() {
            @Override
            public void onQuotesLoaded(List<Quote> quotes) {
                mQuotesLocalDataSource.persist(quotes.get(0)); // TODO for now the remote DS only returns a list with one new element that needs to be persisted
                listener.onQuotesLoaderTaskComplete(quotes);
            }

            @Override
            public void onDataNotAvailable() {
                mQuotesLocalDataSource.getQuotes(language, new QuotesDataSource.GetQuoteCallback() {
                    @Override
                    public void onQuotesLoaded(List<Quote> quotes) {
                        listener.onQuotesLoaderTaskComplete(quotes);
                    }

                    @Override
                    public void onDataNotAvailable() {
                        // TODO error
                    }
                });
            }
        });
    }

    public void addOneTimeShowed(Quote quote) {
        mQuotesLocalDataSource.addOneTimeShowed(quote);
    }
}

File: app/src/main/java/com/sofps/inspirationalquotes/data/source/QuotesRepository.kt
package com.sofps.inspirationalquotes.data.source

import com.sofps.inspirationalquotes.data.source.local.QuotesLocalDataSource
import com.sofps.inspirationalquotes.data.source.remote.QuotesRemoteDataSource
import com.sofps.inspirationalquotes.data.toQuoteDb
import com.sofps.inspirationalquotes.data.toQuoteModel
import com.sofps.inspirationalquotes.model.QuoteModel
import com.sofps.inspirationalquotes.model.ViewState
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.*
import timber.log.Timber

class QuotesRepository(
        private val quotesLocalDataSource: QuotesLocalDataSource,
        private val quotesRemoteDataSource: QuotesRemoteDataSource
) {

    suspend fun getQuotesForLanguage(language: String): Flow<ViewState<List<QuoteModel>>> {
        return flow {
            emit(ViewState.loading())

            try {
                quotesRemoteDataSource.getQuotes(language)
                        .collect { quotes ->
                            Timber.d("quotes=$quotes")
                            quotesLocalDataSource.persist(quotes)
                        }
            } catch (e: Exception) {
                Timber.d("Error when trying to get quotes from remote: $e")
            }

            quotesLocalDataSource.getQuotes(language)
                    .map { quotes ->
                        quotes.map {
                            it.toQuoteModel()
                        }
                    }
                    .collect {
                        emit(ViewState.success(it))
                    }
        }.catch {
            emit(ViewState.error(it.message.orEmpty()))
        }.flowOn(Dispatchers.IO)
    }

    fun addOneTimeShowed(quote: QuoteModel) {
        quotesLocalDataSource.addOneTimeShowed(quote.toQuoteDb())
    }
}
